### DISCUSSION TITLE
Serial build, is it possible?

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#18375] [Q&A General] [answered]
HI All

Ive only ever wanted to build MOOSE in parallel for use on HPC clusters, but now I have a rather unique need to compile in serial. It seems that this isn't currently possible using the offered build scripts.

During the build for PETSC, with no MPI installed I got a message suggesting I set `--with-mpi=0` for the petsc build, however when I do the following; `./scripts/update_and_rebuild_petsc.sh --prefix=/opt/petsc --with-mpi=0`, I get
```
==========================================================================================      Trying to download git://https://gitlab.com/slepc/slepc.git for SLEPC               ====================================================================================================================================================================================      SLEPc examples are available at ${PETSC_DIR}/arch-moose/externalpackages/git.slepc        export SLEPC_DIR=/opt/petsc                                                         ====================================================================================================================================================================================      Trying to download git://https://bitbucket.org/petsc/pkg-metis.git for METIS        ====================================================================================================================================================================================      Configuring METIS with cmake; this may take several minutes                         ====================================================================================================================================================================================      Compiling and installing METIS; this may take several minutes                       ==========================================================================================TESTING: checkDependencies from config.packages.parmetis(config/BuildSystem/config/package*******************************************************************************
         UNABLE to CONFIGURE with GIVEN OPTIONS    (see configure.log for details):
-------------------------------------------------------------------------------
Did not find package MPI needed by parmetis.
Enable the package using --with-mpi
*******************************************************************************

There was an error. Exiting...     
```
Before I go too far down this rabbit hole, does any one know if its currently possible?

**Comment by GiudGiud** [comment]
@fdkong for petsc build

**Comment by fdkong** [comment, marked as answer]
It is possible to build PETSc without MPI even though we do not do that often. You need to shut off any third-party package that requires MPI, for example, parmetis, by doing "--download-xxxx = 0".

But we might not have tests for checking libmesh-and-moose-without-MPI. However, I would be interested  in the results. It is good to try it out. 

Could you please kindly let us what is the motivation here to build packages without MPI?

**Reply by makeclean** [reply]
Sure thing, in this case I've wrapped a geometry using the GMSH python interface, which is driven by OpenMDAO using their pyDOE2 module to drive the exploration of the design space, calling a moose application a bunch of times. This can be done in parallel, and is done by launching your python script with mpi4py. The issue here, is that since the MOOSE application doesnt initialise MPI, there is already an a running MPI initialised, and doesnt seem to play nicely. My runs are short, and thus MOOSE doesn't need to run in parallel.

**Reply by am-tc01** [reply]
@makeclean Just wondering if you could succeed with a serial build of MOOSE?

**Reply by makeclean** [reply]
I think you've jumped the gun here @GiudGiud - no, it failed saying METIS couldnt find MPI even with such arguments.

**Reply by GiudGiud** [reply]
what argument did you use?
can you pls share the build log?

**Comment by gwenchee** [comment]
@makeclean Did you figure this out? I want to do a similar setup to run a MOOSE sub-app with an optimization wrapper.

**Reply by makeclean** [reply]
I did not, if you have any success let us know :)
